### ISSUE TITLE
What all changes should one do to convert a custom environment as a vectorised environment

### ISSUE THREAD

**Issue by muk465** [DLR-RM/stable-baselines3#1089]
I have a custom env and it is made for a single agent below is small snippet:
```

k=self.rad_curvature
u1=self.insertion_depth

#un normalise the action
u2=math.pi*(action[0]+1)
print("action",u2)
xi=self.state[0]

yi=self.state[1]
zi=self.state[2]

qi1=self.state[3]
qi2=self.state[4]
qi3=self.state[5]
qi4=self.state[6]



curr_needle_mtx=self.action_model.homogenous_matrix(xi,yi,zi,qi1,qi2,qi3,qi4)
next_needle_mtx=self.action_model.NeedleModelBicycle(curr_needle_mtx,k,u2,u1)

xnew=next_needle_mtx[0,3]
ynew=next_needle_mtx[1,3]
znew=next_needle_mtx[2,3]

rot_mtx=next_needle_mtx[0:3,0:3]
ar=R.from_matrix(rot_mtx)
Q=ar.as_quat()


self.state_temp[0]=xnew 
self.state_temp[1]=ynew
self.state_temp[2]=znew

self.state_temp[3]=Q[0]
self.state_temp[4]= Q[1]
self.state_temp[5]= Q[2]
self.state_temp[6]= Q[3]

 
print("tempstate: ",self.state_temp)

self.new_state=self.state_temp.copy()

done=False 
self.num_steps+=1

self.reward,done,res,self.num_times_tum_reached=state_reward_validator(self.img_arr,curr_state=self.state_temp,prev_state=self.state,reward=self.reward,num_steps=self.num_steps,steps_to_reset=self.steps_to_reset,done=done,goal_point=self.goal_point,num_times_tum_reached=self.num_times_tum_reached)
```
1)What i want to know is I am using stable baslines 3 and after reading vectorised env doc [](https://stable-baselines3.readthedocs.io/en/master/guide/vec_envs.html)

2)Do i make changes in the actions and rewards like put them in a loop for each env?

3)I also did this after reading stable baseline tutorial on multiprocessing
def make_env(rank, seed=0):
```
    """
    Utility function for multiprocessed env.

    :param env_id: (str) the environment ID
    :param num_env: (int) the number of environments you wish to have in subprocesses
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    """
    def _init():
        env = NeuroRL4(label_name)
        env.seed(seed + rank)
        return env
    set_random_seed(seed)
    return _init

if __name__ == '__main__':
    #env_id = "CartPole-v1"
    num_cpu = 4  # Number of processes to use
    # Create the vectorized environment
    env = DummyVecEnv([make_env(i) for i in range(num_cpu)])

    # Stable Baselines provides you with make_vec_env() helper
    # which does exactly the previous steps for you.
    # You can choose between `DummyVecEnv` (usually faster) and `SubprocVecEnv`
    # env = make_vec_env(env_id, n_envs=num_cpu, seed=0, vec_env_cls=SubprocVecEnv)

    model = DDPG('MlpPolicy', env,train_freq=1, gradient_steps=-1, verbose=1)
    model.learn(total_timesteps=25_000)
```
This above thing is working fine ,so i wanted to confirm is the above thing correct and then what is the difference between this and vectorised env wrappers?

**Comment by qgallouedec** [comment]
Please fill up the issue template completely

**Comment by muk465** [comment]
@qgallouedec can you please elaborate? i have opened issue under custom environments? what else is to be done

**Comment by qgallouedec** [comment]
Exactly. From what I understand, your problem is the vectorization of your custom environment.  So you have to go to New Issue > Custom Gym Environment Issue.  No need to reopen a new issue, just copy and paste and fill the template into this one.

**Comment by qgallouedec** [comment]
> 1)What i want to know is I am using stable baslines 3 and after reading vectorised env doc

Is that a question ?

> 2)Do i make changes in the actions and rewards like put them in a loop for each env?

Not sure to understand your point. From the [doc](https://stable-baselines3.readthedocs.io/en/master/guide/vec_envs.html):

"Because of this, actions passed to the environment are now a vector (of dimension n). It is the same for observations, rewards and end of episode signals (dones)"

You can think of a vectorized environment as an overlay of several environments, to which you send not one action but a list of actions, each of which will be passed to its corresponding environment. Thus, each environment sends you back its new observation (and reward), all stacked in a single array.

> i wanted to confirm is the above thing correct

If it comes from the doc, it has to be correct

> what is the difference between this and vectorised env wrappers?

This? Is your question "_What is the difference between vectorized environment and vectorised env wrapper?_" ?

**Comment by muk465** [comment]
1)see what i basically want to know is i got confused after reading the doc , i clearly understand that actions and rewards become vectors ,do i need to make changes in the environment itself or not becoz in the above link on vectorised environment it seems to me that we need to redefine the step ,reset etc functions to incorporate vectors as input .
2)
```
### Code example
    """
    Utility function for multiprocessed env.

    :param env_id: (str) the environment ID
    :param num_env: (int) the number of environments you wish to have in subprocesses
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    """
    def _init():
        env = NeuroRL4(label_name)
        env.seed(seed + rank)
        return env
    set_random_seed(seed)
    return _init

if __name__ == '__main__':
    
    num_cpu = 4  # Number of processes to use
    # Create the vectorized environment
    env = DummyVecEnv([make_env(i) for i in range(num_cpu)])

    # Stable Baselines provides you with make_vec_env() helper
    # which does exactly the previous steps for you.
    # You can choose between `DummyVecEnv` (usually faster) and `SubprocVecEnv`
    # env = make_vec_env(env_id, n_envs=num_cpu, seed=0, vec_env_cls=SubprocVecEnv)

    model = DDPG('MlpPolicy', env,train_freq=1, gradient_steps=-1, verbose=1)
    model.learn(total_timesteps=25_000)
```
This is a example i modified for the function DummyVecEnv was taken from the example provided by stable baseline itself [](https://colab.research.google.com/github/Stable-Baselines-Team/rl-colab-notebooks/blob/sb3/multiprocessing_rl.ipynb) and seems to work too ,but this doesnt need me to change my environment in any way,so i am confused whats correct approach?

**Comment by qgallouedec** [comment]
No, there is no need to modify the environment. You can find numerous example in the documentation/colab notebooks.

>  in the above link on vectorised environment it seems to me that we need to redefine the step

No. I'm pretty sure you are referring to VecEnvWrapper. Read carefully the doc:
_If you want to alter or augment a VecEnv without redefining it completely (e.g. stack multiple frames, monitor the VecEnv, normalize the observation, …), you can use VecEnvWrapper for that._

**Comment by muk465** [comment]
Thanks @qgallouedec  ,so in my environment i modify my step funtion ?
But   `  env = DummyVecEnv([make_env(i) for i in range(num_cpu)])`
doesnt  DummyVecEnv take one environment at a time sequentially? so why do i need to modify the step function?

**Comment by qgallouedec** [comment]
Please read carefully my answers:

>> [...] we need to redefine the step

> No.

**Comment by muk465** [comment]
after reading the source code for dummyvec env [link](https://stable-baselines3.readthedocs.io/en/master/_modules/stable_baselines3/common/vec_env/dummy_vec_env.html#DummyVecEnv)
1)do i need to add 2 functions step_async and step_wait in my environment?
2)or Do i need to put a loop inside my step in the  env for a vector of action , states,dones and rewards?
eg
```
 for i in range(len(actions)):
u1=actions[i]
states[i]=

```etc.
3)If possible can you provide an example of vectorised env being trained for a custom environment?i cant find examples for a custom vectorised environment .

**Comment by qgallouedec** [comment]
I've already answered twice to 1) and 2) : No.

For 3), let's take PPO as an example. It mainly consists in combining [Using Custom Environments](https://stable-baselines3.readthedocs.io/en/master/guide/custom_env.html#using-custom-environments) and [PPO Example](https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html#example) documentation sections:

```python
import gym

from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env

class CustomEnv(gym.Env):
    ...

# Parallel environments
env = make_vec_env(MyCustomEnv, n_envs=4)

model = PPO("MlpPolicy", env)
model.learn(total_timesteps=25_000)
```

**Comment by muk465** [comment]
Thanks for clearing the confusion ,one last thing i modified make_vec_env as direct function doesnt work for my env as its not registered so doest have a return for gym.make
so i did this 
1)
`def make_env(rank, seed=0):
    """
    Utility function for multiprocessed env.

    :param env_id: (str) the environment ID
    :param num_env: (int) the number of environments you wish to have in subprocesses
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    """
    def _init():
        env = NeuroRL4(label_name)
        env.seed(seed + rank)
        return env
    set_random_seed(seed)
    return _init`

2)so according to your code above we can just pass the parallel env to algo like ppo or ddpg in my case 
and i dont need to do this `env = DummyVecEnv([make_env(i) for i in range(num_env)])`
so what is the  DummyVecEnv doing and why was it used in the sb3 eg and why were they not directly giving parallel env after make_env to the algo in this [link](https://colab.research.google.com/github/Stable-Baselines-Team/rl-colab-notebooks/blob/sb3/multiprocessing_rl.ipynb)

**Comment by qgallouedec** [comment]
> i modified make_vec_env as direct function doesnt work for my env as its not registered

It is not necessary, `make_vec_env` works with unregistered envrionments

**Comment by muk465** [comment]
in my case it was showing this error **'NeuroRL4' object is not callable** where NeuroRL4 is my env name

**Comment by muk465** [comment]
also can you please clear the confusion on the second point?

**Comment by qgallouedec** [comment]
> in my case it was showing this error **'NeuroRL4' object is not callable** where NeuroRL4 is my env name

If `make_vec_env` raises an error with your custom environment, provide a **minimal** and **working** code to reproduce the error. Also, please make sure to go through all the steps explained in the _Custom Gym Environment Issue_.

> so what is the DummyVecEnv doing

`make_vec_env` is an helper function to create vectorized environments. This function returns an instance of  `DummyVecEnv` or `SubprocVecEnv`. The answer was [in the documentation](https://stable-baselines3.readthedocs.io/en/master/common/env_util.html#stable_baselines3.common.env_util.make_vec_env)
